src/blob_format.h

```diff
@@ -61,15 +61,30 @@ struct BlobRecord {
 
 class BlobEncoder {
  public:
-  BlobEncoder(CompressionType compression,
+  BlobEncoder(CompressionType compression, CompressionOptions compression_opt,
               const CompressionDict& compression_dict)
-      : compression_ctx_(compression),
-        compression_info_(compression_opt_, compression_ctx_, compression_dict,
-                          compression, 0 /*sample_for_compression*/) {}
+      : compression_opt_(compression_opt),
+        compression_ctx_(compression),
+        compression_info_(new CompressionInfo(compression_opt, compression_ctx_,
```

Comment:
```suggestion
        compression_info_(new CompressionInfo(compression_opt_, compression_ctx_,
```
Also the compression_dict need to be a member of `BlobEncoder`. This will fix CI.

Reply:
I will try this to fix this later.

I can't reproduce the stack-buffer-underflow with gcc or clang-10 (both with asan). It's a little bit strange. Also when using clang-10, it will throw a rocksdb part error, saying 
```
In file included from /home/hexyoungs/workspace/pingcap/rocksdb/utilities/backupable/backupable_db.cc:19:
/home/hexyoungs/workspace/pingcap/rocksdb/util/channel.h:35:33: error: no matching constructor for initialization of 'std::lock_guard<std::mutex>'
    std::lock_guard<std::mutex> lk(lock_);
```

with following patch(tikv/rocksdb) it will compile, but still will not hit the `stack-buffer-underflow` error in test cases.

```
diff --git a/util/channel.h b/util/channel.h
index 0225482c0..2eaf02d0c 100644
--- a/util/channel.h
+++ b/util/channel.h
@@ -31,7 +31,7 @@ class channel {
     return buffer_.empty() && eof_;
   }

-  size_t size() const {
+  size_t size() {
     std::lock_guard<std::mutex> lk(lock_);
     return buffer_.size();
   }
```

Reply:
reproduced in a VM, ubuntu20 with clang-7, I'll figure out what's happening and fix this tomorrow.

Reply:
trust me, just apply the fix I said..

Reply:
sure, I'll try

Reply:
I don't quite understand this, but it works, awesome!

Reply:
params of CompressionInfo constructor are references, not passing by value.

Reply:
yeah, I got it. The problem is  about the first param. And after playing this around, actually change `compression_opt` to `compression_opt_` is enough to fix the ASAN error. But yeah we better keep `compression_dict` as well.